Define measures

atomczak opened the pull request:
First attempt to redefine IFC measures. The table contains the list of all IFC4.3 measures taken from the IDS project ([units](https://github.com/buildingSMART/IDS/blob/development/Documentation/UserManual/units.md), [data types](https://github.com/buildingSMART/IDS/blob/development/Documentation/ImplementersDocumentation/DataTypes.md)) . 

Column "In IFC5?" defines whether the measure should stay. As of now, 32 are to stay, 19 to be removed, and 41 undecided (most probably will stay).

Additionally, all measures are assigned a unit in which the data should be stored.

The last column (Comments) provides a reason for removal or a decision to be made.

aothms commented on schema/Measures.md:
```diff
@@ -0,0 +1,94 @@
+| IFCPOWERMEASURE                               | Power                                   | Real      | Watt                             | W                   | (2, 1, -3, 0, 0, 0, 0)  | ✔️             | ✔️ (derived)             | ✔️       | ✔️     | ✔️       | ✔️     | 1       |                                                                                    |
+| IFCVOLUMEMEASURE                              | Volume                                  | Real      | Cubic metre                      | m<sup>3</sup>       | (3, 0, 0, 0, 0, 0, 0)   | ✔️             | ❌                        | ✔️       | ✔️     | ✔️       | ✔️     | 1       |                                                                                    |
+| IFCSOUNDPOWERLEVELMEASURE                     | Sound power level                       | Real      | Decibel                          | db                  |                         | ❌              | ❌                        | ❌        | ✔️     | ✔️       | ✔️     | 2       |                                                                                    |
+| IFCPHMEASURE                                  | pH                                      | Real      | pH                               | pH                  |                         | ❌              | ❌                        | ✔️       | ✔️     | ✔️       | ✔️     | 2       | Give it other name (acidity?)                                                      |
```
+1 on acidity

aothms commented:
Currently the measures serve the purpose of property values as well as in schema constructs. Things like positive-length-measure vs non-negative are used extensively in the profile definitions for example. So I agree that for properties+IDS it's a weird and potentially contradictory footgun where we have (measure+constraint)+constraint, but in the schemas it can be a useful shortcut even if they have the same underlying semantics. Of course, the constraints can be moved out of the type declaration into the classes, but that's a lot of repetition. Anyway, just wanted to point out the rationale of why we have so many overlapping measures for length.

I'm also very interested in the intended relationship between measures and units. In IFC4x it's messy with a symbolic coupling between them. The ability to define custom derived units, but not derived measures. It seems this is already miles (or meters; a length quantity) in a better direction.

But I think to properly review this we need to know how these measures intended to be referenced, only in schemas or also in the data? Will IfcDerivedUnit/-Measure keep existing or is this it?

And based on that, I hate to be that guy, but what about https://qudt.org/3.1.0/vocab/quantitykind If we don't want to import everything, maybe just a purely strict subset of qudt would be nice? Just saves us some debating if we can direct to an authoritative source.

atomczak commented:
> Currently the measures serve the purpose of property values as well as in schema constructs. Things like positive-length-measure vs non-negative are used extensively in the profile definitions for example.

I get it, but having so many ways to define lengths leads to inconsistent data. The same goes for IfcText vs IfcLabel properties. I hope we can systematically prevent that.

> And based on that, I hate to be that guy, but what about https://qudt.org/3.1.0/vocab/quantitykind If we don't want to import everything, maybe just a purely strict subset of qudt would be nice? Just saves us some debating if we can direct to an authoritative source.

Agree, we do the same in the bSDD. I wouldn't take all quantity kinds from QUDT as there are 1164, but those that we have should reflect QUDT. This way we can rely on QUDT list of applicable units, definitions etc. Is that what you had in mind @aothms ? Following QUDT, also renamed that table to quantity kinds (not measures).

Since the QUDT doesn't impose any particular unit, and the group agreed in Budapest that IFC should have a unified list of units for storing values, I used SI and common sense to select those. This and other decisions are to be discussed with the community.

Small problem - some of the QUDT quantities don't match IFC/SI. For example, QUDT uses Stokes ST for KinematicViscosity and SI has m2/s. Same goes for decibels. But we could resolve all of those. 

I added two columns mapping to QUDT, one to quantity kind and one to a unit. Also added draft explanation of the assumptions and moved it to docs. Also added a source excel file from which the table was made.

aothms commented:
> Is that what you had in mind @aothms ?

100%

> QUDT uses Stokes ST for KinematicViscosity and SI has m2/s. 

That's just cm2/s vs m2/s, but I agree, I would vote for SI just for the consistency of dealing with derived units.

> should have a unified list of units for storing values

I'm not sure about the *storing*, but the "default coupling" really need to be there.

But I do want to touch upon currencies, counts (and times): 300 trucks of sand vs 300 m3 of sand, 300 usd vs 300 eur, 300 days vs 300 workdays. These things matter in their quantisation and cannot be converted without context (exchange rate, calendar of working days, etc.). I don't think we can say that every non-physical measure degrades to a measureless number. So I would argue that in any case some provisions for units remain necessary in the file scope. Currency as probably the clearest example.

But even for physical ones, similarly, a 65mph road is not just a 104.60736kmh road, it's really akin to an enumeration, so that means if you only support storing values in SI, your users will store that value as `"65mph"` as a string instead of 104.60736kmh. Which means it is less than "7mph" (sorry bad example). It's not a huge problem, because little computation happens based on this number, a bit like how fire rating is a label now, it really functions like an enumeration, but its still ordinal in nature and you break that when wrapping it in a string.

If we make clear choices, like (a) schema of measures with units in the header with a clear mandatory relationships to your table (i.e you cannot deviate from it, but you can specialise it) (b) specify how unit handling ought to happen as a part of composition so that implementation is consistent (c) nevertheless only affect the "property domain" and not the geometry. Then I think the best of both worlds is really attainable.

atomczak commented:
I'm not so sure we need globally defined quantity kinds (previously measures) for ALL of these. First, the list will never be complete. There will always be missing kinds like productivity measured in bricks per area, stripes per kilometre, or Angel's example: "aliens per planet"; missing currencies (think crypto), or missing units (Jin, Tsumbo, Dunam, Fanega...). Second, if the list is too long, users are less likely to use it and will invent new anyway (as we experienced with IFC properties).

`300 trucks of sand vs 300 m3 of sand` - one is an integer, the other one is a number representing volume. If you specify using volumetric, the software can help you convert. If you go for trucks as a unit, don't expect software to convert between types of trucks.
`300 usd vs 300 eur` - if you expect any software to make a conversion you would need to provide a datetime at least.
`300 days vs 300 workdays` - those can be integers with meaning explained in the property definition.

Nice edge case with the 65mph road. I would let users decide if they want to define it as a non-convertible string enumeration or a speed measure. With fewer options, it is easier to make that decision :)

aothms commented:
Just one comment:

>  300 usd vs 300 eur - if you expect any software to make a conversion you would need to provide a datetime at least.

I don't care here about conversion, I care about the users ability to write it down in a somewhat structured way.
